File: src/main/java/net/thedragonteam/armorplus/compat/jei/workbench/WBCategory.java
/*
 * Copyright (c) TheDragonTeam 2016.
 */

package net.thedragonteam.armorplus.compat.jei.workbench;

import mezz.jei.api.gui.ICraftingGridHelper;
import mezz.jei.api.gui.IDrawable;
import mezz.jei.api.gui.IGuiItemStackGroup;
import mezz.jei.api.gui.IRecipeLayout;
import mezz.jei.api.ingredients.IIngredients;
import mezz.jei.api.recipe.BlankRecipeCategory;
import mezz.jei.api.recipe.wrapper.ICraftingRecipeWrapper;
import mezz.jei.api.recipe.wrapper.ICustomCraftingRecipeWrapper;
import mezz.jei.api.recipe.wrapper.IShapedCraftingRecipeWrapper;
import mezz.jei.util.Translator;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ResourceLocation;
import net.thedragonteam.armorplus.api.Constants;
import net.thedragonteam.armorplus.compat.jei.ArmorPlusPlugin;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.List;

public class WBCategory extends BlankRecipeCategory<ICraftingRecipeWrapper> {

    public static final int width = 116;
    public static final int height = 54;
    private static final int OUTPUT_SLOT = 0;
    private static final int INPUT_SLOT = 1;
    @Nonnull
    private final IDrawable background;
    @Nonnull
    private final String localizedName;
    @Nonnull
    private final ICraftingGridHelper craftingGridHelper;

    public WBCategory() {
        ResourceLocation location = new ResourceLocation("armorplus", "textures/gui/container/gui_workbench.png");
        background = ArmorPlusPlugin.jeiHelper.getGuiHelper().createDrawable(location, 29, 16, width, height);
        localizedName = Translator.translateToLocal("gui.jei.category.workbench");
        craftingGridHelper = ArmorPlusPlugin.jeiHelper.getGuiHelper().createCraftingGridHelper(INPUT_SLOT, OUTPUT_SLOT);
    }

    @Nonnull
    @Override
    public String getUid() {
        return Constants.Compat.JEI_CATEGORY_WORKBENCH;
    }

    @Nonnull
    @Override
    public String getTitle() {
        return localizedName;
    }

    @Nonnull
    @Override
    public IDrawable getBackground() {
        return background;
    }

    @Nullable
    @Override
    public IDrawable getIcon() {
        return null;
    }

    @Override
    public void setRecipe(IRecipeLayout recipeLayout, ICraftingRecipeWrapper recipeWrapper, IIngredients ingredients) {
        IGuiItemStackGroup guiItemStacks = recipeLayout.getItemStacks();

        guiItemStacks.init(OUTPUT_SLOT, false, 94, 18);

        for (int y = 0; y < 3; ++y) {
            for (int x = 0; x < 3; ++x) {
                int index = INPUT_SLOT + x + (y * 3);
                guiItemStacks.init(index, true, x * 18, y * 18);
            }
        }

        if (recipeWrapper instanceof ICustomCraftingRecipeWrapper) {
            ICustomCraftingRecipeWrapper customWrapper = (ICustomCraftingRecipeWrapper) recipeWrapper;
            customWrapper.setRecipe(recipeLayout, ingredients);
            return;
        }

        List<List<ItemStack>> inputs = ingredients.getInputs(ItemStack.class);
        List<List<ItemStack>> outputs = ingredients.getOutputs(ItemStack.class);

        if (recipeWrapper instanceof IShapedCraftingRecipeWrapper) {
            IShapedCraftingRecipeWrapper wrapper = (IShapedCraftingRecipeWrapper) recipeWrapper;
            craftingGridHelper.setInputs(guiItemStacks, inputs, wrapper.getWidth(), wrapper.getHeight());
        } else {
            craftingGridHelper.setInputs(guiItemStacks, inputs);
            recipeLayout.setShapeless();
        }
        guiItemStacks.set(OUTPUT_SLOT, outputs.get(0));
    }
}

File: src/main/java/net/thedragonteam/armorplus/compat/jei/workbench/WBCategory.kt
/*
 * Copyright (c) TheDragonTeam 2016-2017.
 */

package net.thedragonteam.armorplus.compat.jei.workbench

import mezz.jei.api.gui.ICraftingGridHelper
import mezz.jei.api.gui.IDrawable
import mezz.jei.api.gui.IRecipeLayout
import mezz.jei.api.ingredients.IIngredients
import mezz.jei.api.recipe.BlankRecipeCategory
import mezz.jei.api.recipe.IRecipeWrapper
import mezz.jei.api.recipe.wrapper.ICustomCraftingRecipeWrapper
import mezz.jei.api.recipe.wrapper.IShapedCraftingRecipeWrapper
import mezz.jei.util.Translator
import net.minecraft.item.ItemStack
import net.minecraft.util.ResourceLocation
import net.thedragonteam.armorplus.api.Constants
import net.thedragonteam.armorplus.compat.jei.ArmorPlusPlugin

class WBCategory : BlankRecipeCategory<IRecipeWrapper>() {
    private val background: IDrawable
    private val localizedName: String = Translator.translateToLocal("gui.jei.category.workbench")
    private val craftingGridHelper: ICraftingGridHelper

    init {
        val location = ResourceLocation("armorplus", "textures/gui/container/gui_workbench.png")
        background = ArmorPlusPlugin.jeiHelper.guiHelper.createDrawable(location, 29, 16, width, height)
        craftingGridHelper = ArmorPlusPlugin.jeiHelper.guiHelper.createCraftingGridHelper(INPUT_SLOT, OUTPUT_SLOT)
    }

    override fun getUid(): String = Constants.Compat.JEI_CATEGORY_WORKBENCH

    override fun getTitle(): String = localizedName

    override fun getBackground(): IDrawable = background

    override fun getIcon(): IDrawable? = null

    override fun setRecipe(recipeLayout: IRecipeLayout, recipeWrapper: IRecipeWrapper, ingredients: IIngredients) {
        val guiItemStacks = recipeLayout.itemStacks

        guiItemStacks.init(OUTPUT_SLOT, false, 94, 18)

        (0.rangeTo(2)).forEach { y ->
            (0.rangeTo(2)).forEach { x ->
                val index = INPUT_SLOT + x + y * 3
                guiItemStacks.init(index, true, x * 18, y * 18)
            }
        }

        if (recipeWrapper is ICustomCraftingRecipeWrapper) {
            recipeWrapper.setRecipe(recipeLayout, ingredients)
            return
        }

        val inputs = ingredients.getInputs(ItemStack::class.java)
        val outputs = ingredients.getOutputs(ItemStack::class.java)

        when (recipeWrapper) {
            is IShapedCraftingRecipeWrapper -> craftingGridHelper.setInputs(guiItemStacks, inputs, recipeWrapper.width, recipeWrapper.height)
            else -> {
                craftingGridHelper.setInputs(guiItemStacks, inputs)
                recipeLayout.setShapeless()
            }
        }
        guiItemStacks.set(OUTPUT_SLOT, outputs[0])
    }

    companion object {

        val width = 116
        val height = 54
        private val OUTPUT_SLOT = 0
        private val INPUT_SLOT = 1
    }
}
